### PR TITLE
Proofs of the Binomial Theorem for `(Commutative)Semiring` [supersedes #1287]

### PR DESCRIPTION
**Pull request by jamesmckinna**
This PR is ~~draft, for discussion~~ ready for review. A major overhaul of @uzytkownik 's #1287, but preserving the basic structure of his argument. 

It depends on the following ancillary PRs for completion (I've tried to keep the branch/commit structure clean):
- [x] #1923, to establish the `top` view of `Fin (suc n)`; UPDATED ~~have now removed dependency on this PR/can be~~ has now been reconfigured to use the ~~revised~~ merged version of it; 
- [x] #1926 for the required arithmetic of `_C_`; 
- [x] #1927 for the ancillary algebraic lemmas (phrased for `Semiring`s but could be weakened; see also issue #1917)

### CODE REVIEW

**Comment by jamesmckinna** [comment]
**TEMPORARY** commit of a copy of `Data.Fin.Relation.Unary.Top` for the sake of testing. Should revert this before merging.

File: src/Algebra/Properties/Semiring/Binomial.agda
Comment (Taneb) [inline src/Algebra/Properties/Semiring/Binomial.agda:156]: Why for `x*lemma` do you have one module taking `n` and `j` simultaneously, but for `y*lemma` you have a module parameterized on `n` with then `j` as an argument to the lemma? If `j` was a module parameter you could define `i` as a local private definition in the module

File: src/Algebra/Properties/Semiring/Binomial.agda
Comment (Taneb) [inline src/Algebra/Properties/Semiring/Binomial.agda:206]: This name is both not particularly informative and pretty damn hard to read (and type). Maybe a different name is in order?

Comment (Taneb) [reply src/Algebra/Properties/Semiring/Binomial.agda:206]: I think that works

**Review by gallais — APPROVED** [review]

File: src/Data/Fin/Relation/Unary/Top.agda
Comment (Taneb) [inline src/Data/Fin/Relation/Unary/Top.agda:81]: This comment is maybe behind the fashions of the library. I think replacing "`--without-K`" with the more prosey "without K" should be good enough

**Comment by MatthewDaggitt** [comment]
Blocked on #1923

**Comment by jamesmckinna** [comment]
@MatthewDaggitt sorry if the commentary history has got too mangled to be readable. The commute I made yesterday during AIM mean that this PR is now no longer dependent on the 'full' version of #1923 ... so I think the 'blocked-on-issue' label is (now) unwarranted...

... I think that #1923  can now be considered separately. It's certainly a much more articulated version of what is now used in the present PR, and potentially has implications for deprecating some existing definitions in `Data.Fin.Base`.

Or did you have something else in mind when adding the label?

**Comment by jamesmckinna** [comment]
I've now revised #1923 (to its betterment, I hope ;-)) in the light of earlier comments, and so will now turn to the possible knock-on effects for this PR. So happy to continue with blocking on the merge of #1923, after which I hope this will be straightforward...

**Comment by jamesmckinna** [comment]
Now returning to this after the merge of #1923...

File: src/Algebra/Properties/Semiring/Binomial.agda
Comment (MatthewDaggitt) [inline src/Algebra/Properties/Semiring/Binomial.agda:42]: While this is really clever structure to the file, and I like the names it generates, unfortunately it's incredibly difficult to work out what the parameterisation of each lemma actually is. For example the crucial commutativity assumption for `theorem` actually occurs 150 lines earlier. I think we need to flatten out most of the module structure....

Is that okay?

Comment (jamesmckinna) [reply src/Algebra/Properties/Semiring/Binomial.agda:42]: Ouch. Probably (of course), but ... fiddly? I'll try to look at it over the weekend.

Comment (MatthewDaggitt) [reply src/Algebra/Properties/Semiring/Binomial.agda:42]: I'll have a look at this now.

**Comment by MatthewDaggitt** [comment]
Had a go at flattening it. @jamesmckinna if it's okay with you I'll merge it in now?

File: src/Algebra/Properties/Semiring/Binomial.agda
Comment (jamesmckinna) [inline src/Algebra/Properties/Semiring/Binomial.agda:44]: 'implicit' here?

File: src/Algebra/Properties/Semiring/Binomial.agda
Comment (jamesmckinna) [inline src/Algebra/Properties/Semiring/Binomial.agda:45]: and 'explicit' here?

**Comment by jamesmckinna** [comment]
Modulo my comments?

File: src/Algebra/Properties/Semiring/Binomial.agda
Comment (jamesmckinna) [inline src/Algebra/Properties/Semiring/Binomial.agda:32]: Explicit or implicit for `x` and `y`?

Comment (jamesmckinna) [reply src/Algebra/Properties/Semiring/Binomial.agda:32]: Ah, no... unless we are willing to make the definition in `CommutativeSemiring` slightly clunkier. Live and learn!

Comment (MatthewDaggitt) [reply src/Algebra/Properties/Semiring/Binomial.agda:32]: Yup I think better to leave these explicit :+1:

**Pull request merged**
